sol/main.c

```diff
@@ -0,0 +1,478 @@
+		req_queue->tb_credits = req_queue->tb_size_credits;
+
+	req_queue->spare_cycles = avail_cycles -
+		(avail_bytes * req_queue->cycles_per_byte);
+	req_queue->time_cpu_cycles = curr_cycles;
+}
+
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	/* Need to include Ethernet frame overhead (preamble, gap, etc.) */
+	uint32_t pkt_len = pkt->pkt_len + RTE_SCHED_FRAME_OVERHEAD_DEFAULT;
+	if (pkt_len > req_queue->tb_credits)
+		return false;
+	req_queue->tb_credits -= pkt_len;
+	return true;
+}
+
+static void
+dequeue_reqs(struct sol_config *sol_conf, uint8_t tx_port)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct priority_req *entry, *next;
+
+	struct rte_mbuf *pkts_out[sol_conf->deq_burst_size];
+	uint32_t nb_pkts_out = 0;
+	uint16_t total_sent = 0;
+
+	/* Get an up-to-date view of our credits. */
+	credits_update(req_queue);
+
+	list_for_each_entry_safe(entry, next, &req_queue->head, list) {
+		struct rte_mbuf *pkt = entry->pkt;
+
+		if (!credits_check(req_queue, pkt)) {
+			/*
+			 * XXX When under an attack, we may not want to
+			 * log this because it could become expensive.
+			 */
+			RTE_LOG(NOTICE, GATEKEEPER,
+				"sol: out of request bandwidth\n");
+			goto out;
+		}
+
+		if (req_queue->len == 1 || (entry->priority != next->priority))
+			req_queue->priorities[entry->priority] = NULL;
+		list_del(&entry->list);
+		mb_free_entry(&sol_conf->mb, entry);
+		req_queue->len--;
+
+		pkts_out[nb_pkts_out++] = pkt;
+
+		if (nb_pkts_out >= sol_conf->deq_burst_size)
+			break;
+	}
+
+out:
+	if (list_empty(&req_queue->head)) {
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+	} else {
+		struct priority_req *first = list_first_entry(&req_queue->head,
+			struct priority_req, list);
+		req_queue->highest_priority = first->priority;
+	}
+
+	/* We cannot drop the packets, so re-send. */
+	while (nb_pkts_out > 0) {
+		uint16_t sent = rte_eth_tx_burst(tx_port,
+			sol_conf->tx_queue_back,
+			pkts_out + total_sent, nb_pkts_out);
+		total_sent += sent;
+		nb_pkts_out -= sent;
+	}
+}
+
+/*
+ * @sol_conf is allocated using rte_calloc(), so initializations
+ * to 0 are not strictly necessary in this function.
+ */
+static int
+req_queue_init(struct sol_config *sol_conf)
+{
+	struct req_queue *req_queue = &sol_conf->req_queue;
+	struct rte_eth_link link;
+	uint64_t link_speed_bytes;
+
+	req_queue->len = 0;
+	req_queue->highest_priority = 0;
+	req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+
+	/* Calculate link speed in bytes per second. */
+	/* XXX Make work for bonded interface. */
+	rte_eth_link_get(sol_conf->net->back.id, &link);
+	if (link.link_speed == ETH_SPEED_NUM_NONE) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"sol: back interface link speed is undefined\n");
+		return -1;
+	}
+	link_speed_bytes = ((uint64_t)link.link_speed * 1000 * 1000 / 8);
+	req_queue->cycles_per_byte =
+		rte_get_tsc_hz() / (sol_conf->req_bw_rate * link_speed_bytes);
```

Comment:
1. We'll still need `spare_cycles`, but there's a small trick to eliminate it as a field: decrementing the spare cycles from `time_cpu_cycles` while updating `time_cpu_cycles`.

2. Given that we need the spare cycles, `rte_reciprocal_*()` is not that useful because it implies two multiplications: one to obtain bytes from cycles, and another to obtain the spare cycles. We can drop `rte_reciprocal_*()` and use `ldiv()` to do both at the cost of a division.

3. Since we're going to save `cycles_per_byte` (the double) as a fraction, suggestion for field names: `cycles_per_byte_a` and `cycles_per_byte_b`.